### PR TITLE
[Hackathon] Merge and tag pt. I

### PR DESCRIPTION
**Pull request by marcin-rzeznicki**
This is the first part of creating the `merge and tag` functionality.
This PR does not do much when it comes to implementing the said functionality (this will be the part II), but it will help the reviewers to see through the rest by pushing some mundane stuff out of the main PR (part II). In this PR:
- The stackage snapshot is pushed to the latest version with GHC `8.8.4`
- HLint's suggestions are (sparingly) applied,
- broken tests fixed (something deadlocked after the update),
- general refactorings (e.g. new functions around `ApprovalType` that were spreading throughout the code)
- formatting,
- working HLS cradle is added,

### CODE REVIEW

**Comment by marcin-rzeznicki** [comment]
There is something rotten in the state of CI

```
Cloning into 'hoff'... 00:01
Warning: Permanently added 'github.com,140.82.121.4' (RSA) to the list of known hosts. 00:01
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ 00:01
@         WARNING: UNPROTECTED PRIVATE KEY FILE!          @ 00:01
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ 00:01
Permissions 0644 for '/home/semaphore/.ssh/id_ed25519' are too open. 00:01
It is required that your private key files are NOT accessible by others. 00:01
This private key will be ignored. 00:01
Load key "/home/semaphore/.ssh/id_ed25519": bad permissions 00:01
git@github.com: Permission denied (publickey). 00:01
fatal: Could not read from remote repository.
```

File: src/Logic.hs [hunk 1]
```diff
@@ -567,7 +570,7 @@ provideFeedback :: ProjectState -> Action ProjectState
 provideFeedback state
   = foldM (flip leaveFeedback) state
   $ filter (Pr.needsFeedback . snd)
-  $ fmap (\(key, pr) -> (PullRequestId key, pr))
+  $ fmap (first PullRequestId)
```
Comment (fatho) [inline src/Logic.hs:573]: Nit: the explicit lambda was somewhat more informative

**Review by fatho — CHANGES_REQUESTED** [review]
Generally LGTM, but not a big fan of the import formatting changes (for the reasons below).

File: src/Logic.hs [hunk 2]
```diff
@@ -32,47 +32,55 @@ module Logic
...
+                                                 PullRequestStatus (..))
+import           Types                          (PullRequestId (..),
+                                                 Username (..))
+
+import qualified Configuration                  as Config
 import qualified Git
 import qualified GithubApi
-import qualified Project as Pr
-import qualified Configuration as Config
+import qualified Project                        as Pr
```
[the remark this answers is not in the document]
Comment (marcin-rzeznicki) [reply src/Logic.hs:75]: Ah, I can try but I just used HLS formatter plugin for sorting and formatting those, I can probably look for the settings that do not add padding

File: src/Logic.hs (hunk 1 above)
Comment (marcin-rzeznicki) [reply src/Logic.hs:573]: I beg to differ - I am not really fond of this _temp_ variables with hastily made-up names :smile:

**Comment by marcin-rzeznicki** [comment]
Reformatted and added `stylish-haskell` config

**Comment by marcin-rzeznicki** [comment]
Ah, let me rebase on master to get rid of the CI errors

File: src/WebInterface.hs [hunk 3]
```diff
@@ -240,15 +240,13 @@ viewPullRequestWithApproval info prId pullRequest = do
...
-              MergeAndDeploy -> "merge and deploy"
-        toHtml $ format "Approved for {} by " [approvedAction :: String]
+        let approvedAction = Project.displayApproval approvalType
+        toHtml $ format "Approved for {} by " [approvedAction]
         -- TODO: Link to approval comment, not just username.
         let url = Text.append "https://github.com/" username
         a ! href (toValue url) $ toHtml username
     Nothing ->
-      fail $
+      error $
```
Comment (fatho) [inline src/WebInterface.hs:249]: I suppose this is `MonadFail` (or the absence thereof for the `Html` monad) related?

File: tests/ServerSpec.hs [hunk 4]
```diff
@@ -132,16 +130,9 @@ withServer body = do
...
-
-  -- Stop the server by killing the thread. This will not immediately stop the
-  -- server, but we do need to ensure that the port is available after we exit
-  -- the function, so the next test can use it, so also wait for the thread to
-  -- finish after killing it.
-  killThread $ asyncThreadId serverAsync
-  _ <- waitCatch serverAsync
+  withAsync runServer $ \_ -> do
+    blockUntilReady
+    body ghQueue
```
Comment (fatho) [inline tests/ServerSpec.hs:135]: Nice, much simpler :+1:

File: tests/ServerSpec.hs [hunk 5]
```diff
@@ -246,7 +237,7 @@ serverSpec = do
         -- First send an invalid webhook with a bad payload.
         let badPayload = "this is definitely not valid json"
         badResponse <- httpPostGithubEvent "/hook/github" "status" badPayload
-        Http.getResponseStatus badResponse `shouldBe` internalServerError500
+        Http.getResponseStatus badResponse `shouldBe` badRequest400
```
Comment (fatho) [inline tests/ServerSpec.hs:240]: Is this caused by a version upgrade of one of our dependencies?

**Review by fatho — APPROVED** [review]
LGTM, just some questions:

File: tests/ServerSpec.hs (hunk 5 above)
Comment (marcin-rzeznicki) [reply tests/ServerSpec.hs:240]: Probably `scotty` at least I did not change anything related. It's what happens on unparseable (or undecodable) JSON

**Comment by marcin-rzeznicki** [comment]
OK, I'll merge this one and onto the final PR

@OpsBotPrime merge

**Comment by OpsBotPrime** [comment]
Pull request approved for merge by @marcin-rzeznicki, rebasing now.

**Comment by OpsBotPrime** [comment]
Rebased as c66b0066063be99bb734313202474e9ff3784d97, waiting for CI …

**Pull request merged**